Changes for 3.0.0-preview1

roji opened the pull request:
This makes the provider compile against the recently-released EF Core 3.0.0-preview.18572.1. Some minor changes were necessary, and some tests were disabled because of EF Core-side changes (especially https://github.com/aspnet/EntityFrameworkCore/issues/14159).

All tests pass, once this is merged I'll release our own 3.0.0-preview1.

austindrenski commented on test/EFCore.PG.FunctionalTests/Query/RangeQueryNpgsqlTest.cs:
```diff
@@ -498,7 +498,7 @@ public void UserDefinedSchemaQualified()
             {
                 var e = context.RangeTestEntities.Single(x => x.SchemaRange == NpgsqlRange<double>.Parse("(0,10)"));
-                AssertContainsSql("WHERE x.\"SchemaRange\" = @__Parse_0");
+                AssertContainsSql("WHERE x.\"SchemaRange\" = '(0,10)'::test.\"Schema_Range\"");
```
That looks great!

roji replied:
Yeah I was impressed too :) It's a bit mind-boggling to have an ORM that's generic and customizable enough to produce this kind of literal...

austindrenski commented on test/EFCore.PG.FunctionalTests/Query/NetworkQueryNpgsqlTest.cs:
```diff
@@ -80,7 +80,7 @@ public void IPAddress_inet_parse_column()
                            .ToArray();
 
-                AssertContainsSql("WHERE x.\"Inet\" = CAST(x.\"TextInet\" AS inet)");
+                AssertContainsSql("WHERE (x.\"Inet\" = CAST(x.\"TextInet\" AS inet))");
```
What changed to add these now?

roji replied:
The full WHERE clause in 3.0 is `WHERE (x."Inet" = CAST(x."TextInet" AS inet)) OR (x."Inet" IS NULL AND x."TextInet" IS NULL)`. IIRC the null checking clause (after the OR) is new, so the first clause is enclosed in parentheses...

austindrenski replied:
Oh... Did any of the configurations change to enable/disable use relational nulls?

austindrenski replied:
Also, if that's the full `WHERE` now, could you update these tests to include it in the assert? (That way if anything changes again...)

roji replied:
As a general rule, the SQL assertions currently only assert the minimal needed to know that on our side the translation occurred correctly, rather than include more SQL that isn't related to our translation (like the 2nd null-checking clause). This comes from my experience in tracking upstream EF Core, where the SQL generated changes quite frequently in small ways, leading to many uninteresting test failures on our side...

austindrenski replied:
Hmm... That's a fair point, considering that these tests only need to verify expression translation, not the relational null handling bit. Consider this resolved, thanks!

austindrenski commented on Directory.Build.props:
```diff
@@ -2,7 +2,7 @@
     <NpgsqlVersion>4.0.4</NpgsqlVersion>
-    <EFCoreVersion>2.2.0</EFCoreVersion>
+    <EFCoreVersion>3.0.0-preview.18572.1</EFCoreVersion>
     <MicrosoftExtensionsVersion>$(EFCoreVersion)</MicrosoftExtensionsVersion>
```
Actually, that package only goes up to 2.1.1 before the 2.2 previews start (whereas EF Core went to 2.1.4), so it makes sense that these are aligned for the preview, but we should keep an eye on it for the hotfix branches.

roji commented on src/EFCore.PG/Query/EvaluatableExpressionFilters/Internal/NpgsqlFullTextSearchEvaluatableExpressionFilter.cs:
```diff
@@ -2,8 +2,8 @@
 using System.Reflection;
 using JetBrains.Annotations;
 using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query.Internal;
```
That's the way things are at the moment... Opened https://github.com/aspnet/EntityFrameworkCore/issues/14170 to make sure this is intended etc.

roji commented:
@austindrenski addressed your comments, please take another look.

@YohDeadfall let us know if you want to take a look, it would be good to release 3.0.0-preview1 soon.

roji commented:
Crap, the build is failing on the plugin tests, forgot to look at those :)

austindrenski commented on src/EFCore.PG/Extensions/NpgsqlServiceCollectionExtensions.cs:
```diff
@@ -5,6 +5,7 @@
 using Microsoft.EntityFrameworkCore.Query;
 using Microsoft.EntityFrameworkCore.Query.ExpressionTranslators;
 using Microsoft.EntityFrameworkCore.Query.ExpressionVisitors;
+using Microsoft.EntityFrameworkCore.Query.Internal;
```
I'm not seeing any changes that use this... did something move upstream, or was this unintentional?

austindrenski replied:
Oh I see, which also explains why Relinq requires the alias just below this. Thanks for clarifying!

austindrenski commented on test/EFCore.PG.FunctionalTests/BuiltInDataTypesNpgsqlTest.cs:
```diff
@@ -108,7 +108,7 @@ public virtual void Can_query_using_any_mapped_data_type()
                         StringAsText = "Gumball Rules!",
                         StringAsVarchar = "Gumball Rules OK",
-                        CharAsChar1 = 'f',
+                        // CharAsChar1 = 'f',
```
Could you add a `TODO` to this file, just to make sure we circle back?

roji replied:
Sure, although I'm not sure that's what's going to make us remember :)

roji commented:
@austindrenski you can take another look

austindrenski commented on src/EFCore.PG/Query/EvaluatableExpressionFilters/Internal/NpgsqlNodaTimeEvaluatableExpressionFilter.cs:
```diff
@@ -24,56 +24,22 @@
+                ||
+                (expression is MemberExpression memberExpr &&
+                memberExpr.Member.DeclaringType?.FullName == "NodaTime.SystemClock" &&
+                memberExpr.Member.Name == "Instance"));
```
It's out of scope for this PR, but is there a reason we aren't storing static fields of the `MethodInfo` and `PropertyInfo` for these comparisons?

roji replied:
Yep, there is :) These are NodaTime symbols. The NodaTime plugin is the one actually referencing the NodaTime assembly, but since there's no plugin-based evaluatable filter option, we have to do this in the core which doesn't/shouldn't reference NodaTime. See https://github.com/aspnet/EntityFrameworkCore/issues/13454 (this is why we have the TODO above.

YohDeadfall commented on test/EFCore.PG.FunctionalTests/BuiltInDataTypesNpgsqlTest.cs:
```diff
@@ -212,8 +212,8 @@ public virtual void Can_query_using_any_mapped_data_type()
-                var param21a = 'f';
-                Assert.Same(entity, context.Set<MappedNullableDataTypes>().Single(e => e.Int == 999 && e.CharAsChar1 == param21a));
+                // var param21a = 'f';
+                // Assert.Same(entity, context.Set<MappedNullableDataTypes>().Single(e => e.Int == 999 && e.CharAsChar1 == param21a));
```
Maybe TODO here?

YohDeadfall commented on test/EFCore.PG.FunctionalTests/BuiltInDataTypesNpgsqlTest.cs:
```diff
@@ -366,8 +366,8 @@ public virtual void Can_query_using_any_mapped_data_types_with_nulls()
-                char? param21a = null;
-                Assert.Same(entity, context.Set<MappedNullableDataTypes>().Single(e => e.Int == 911 && e.CharAsChar1 == param21a));
+                // char? param21a = null;
+                // Assert.Same(entity, context.Set<MappedNullableDataTypes>().Single(e => e.Int == 911 && e.CharAsChar1 == param21a));
```
And here.

roji commented:
Added TODOs everywhere where stuff is commented out.
